### DISCUSSION TITLE
HDR handling in WebGPU & WebCodec intergation

### DISCUSSION THREAD

**Discussion by kedanielwu** [gpuweb/gpuweb#4384] [Q&A] [answered]
I am working on a HDR -> SDR video rendering project in browser environment, the HDR spec is HDR10 compitable but also come with a custom tone-mapping process, the video comes with yuv420p10le pixel format, bt2020 color primaries, PQ transfer, and dynamic metadata in SEI payload. Therefore what I need is using the dynamic metadata to perform custom tone-mapping on RGB value in shader to get a 'perhaps' better SDR image quality over native HDR10 -> SDR tone mapping.

I am new to GPU world, so my question is, is the current implementation of WebGPU and Webcodecs in Chrome capable for me to such task?

What I am currently doing is 
step1: using webcodec to get VideoFrame (unfortunately yuv420p10e is not well supported and I can't use copyTo to get YUV raw data since VideoFrame.format returns null)
step2: import VideoFrame as external texture and load into WebGPU shader.
step3: use textureSampleBaseClampToEdge to get pixel color.

I think my custom tone-mapping process should happens after step3, but I am not so sure about what textureSampleBaseClampToEdge returns to me, and the rendered image looks 'grayish' / 'washed out'. After look through Chromium's source code, it seems currently Dawn will perform gamma correction and YUVtoSRGB conversion with bt709, I am kind lost here, should my custom tone-mapping run on top of the result of textureSampleBaseClampToEdge? Or is there any other approach?

**Comment by Kangz** [comment]
I don't know much about colorspaces so take this with a grain of salt. My understanding is that WebGPU *should* do the conversion for you from the `bt2020` color primaries of the video, to whatever you passed to `GPUExternalTextureDescriptor.predefinedColorspace`. However Chromium might not implement all of that logic correctly yet especially since you say that "yuv420p10e is not well supported yet". Might be worth filing an issue on crbug.com?

**Reply by kedanielwu** [reply]
thanks for the advice. what I mean for "yuv420p10e is not well supported yet" is Webcodec simply not return VideoFrame.format for yuv420p10le and I think 10bit / HDR is not fully supported by webcodec (although you can get correct color information from VideoFrame object), so it is expected.

what I really want to know from WebGPU side is, given the current status of Webcodec intergration with WebGPU, is it possible to do custom YUV -> RGB conversion and custom tone-mapping using VideoFrame as external texture (i.e load via importExternalTexture()), if the given VideoFrame is 10bit.

Also I think currently Chromium only support colorSpace: 'srgb' for importExternalTexture with VideoFrame object? using display-p3 will throw unsupported error, and HDR related color spaces are not yet added. So based on your reply, if I understand correctly, WebGPU will automatically perform bt2020 to srgb (rec709?) conversion.

**Reply by kainino0x** [reply]
> is it possible to do custom YUV -> RGB conversion
This is not possible. GPUExternalTexture encapsulates the conversion to RGB and doesn't provide access to YUV. There have been some feature requests for this, but we don't have one filed on this repository yet. If you'd like to file one, describing a concrete use-case would be very helpful.

However we do (try to) strongly define what the RGB results will be, and you can do further conversion work on the RGB values.
In _theory_ the precision of a 10-bit input should be preserved, but I am definitely not sure if this is true in Chrome. In particular, if we are hitting the 1-copy path (convert to RGB before calling the shader) then we may be converting to RGB8. Please file on crbug.com/new if you find that we are dropping to 8-bit precision.

**Reply by kainino0x** [reply]
One of the past discussions was https://github.com/gpuweb/gpuweb/pull/1681

**Reply by shaoboyan** [reply]
>using display-p3 will throw unsupported error, 

According to [issue 1330250](https://bugs.chromium.org/p/chromium/issues/detail?id=1330250), this check should be removed. And `display-p3` should be supported. @jchen10  told me if this could achieve, you could apply HDR tone-mapping based on the transfer results in shader.

> However Chromium might not implement all of that logic correctly yet especially since you say that "yuv420p10e is not well supported yet". Might be worth filing an issue on crbug.com?

(From chromium impl part).
Yes, the HDR support is still missing in current GPUExternalTexture(0-copy path) and I think [issue 1330230](https://bugs.chromium.org/p/chromium/issues/detail?id=1330230) is tracking on this. And for 1-copy path, we will fallback to 8-bit precision(And it only handles [predefined color space conversion](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/canvas/predefined_color_space.cc;drc=445f8cfdbb63b29ce44f728f6b5015cf18df5534;l=14)).

> the HDR spec is HDR10 compitable but also come with a custom tone-mapping process, the video comes with yuv420p10le pixel format, bt2020 color primaries, PQ transfer, and dynamic metadata in SEI payload. 

But I'm not sure whether this could be achieved. I don't think we have API or parameters to accept custom tone-mapping process. So for current `textureSampleBaseClampToEdge ()` I don't think it is the proper choice.

>One of the past discussions was https://github.com/gpuweb/gpuweb/pull/1681

And yes, I think one candidate solution is to expose the raw data in shader and user could upload their own metadata through buffer and handle it. But only the luminance layer is not enough.
Or 
we could extend current `GPUExternalTextureDescriptor` to allow the customer defined color space conversion metas.

**Reply by kedanielwu** [reply]
Thank you all for such detailed information, I tried read through the code path of Chromium for importing external texture for HDR content, seems we are losing prcesion along the way as for now (10bit -> 8bit, rec2020 -> sRGB, etc..)

> But I'm not sure whether this could be achieved. I don't think we have API or parameters to accept custom tone-mapping process. So for current textureSampleBaseClampToEdge () I don't think it is the proper choice.

> you could apply HDR tone-mapping based on the transfer results in shader.

If `textureSampleBaseClampToEdge` is not proper choice for now, is there any suggestion on how to implement this feature? One way I could think about is getting YUV raw data from Webcodec(although not possoble right now) -> copy YUV as data buffer to WebGPU storage -> perform conversion and tone-mapping -> output srgb to canvas, I am wondering if this is technically correct path.

Two issue I can think about in production are:
(1) frame buffer copy in JS is costly
(2) although I am working with YUV, I think as developer I still missing some information (like display setup) to correctly perform tone-mapping.

> And for 1-copy path, we will fallback to 8-bit precision(And it only handles [predefined color space conversion](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/canvas/predefined_color_space.cc;drc=445f8cfdbb63b29ce44f728f6b5015cf18df5534;l=14)

So If I understanded correctly, currently 1-copy path will drop the precision to 8bit then perform color-space conversion. I do see `bt2020` & `bt2100-pq` in `V8PredefinedColorSpace`, but not yet exposed to JS WebGPU context therefore colorSpace is currently either `srgb` or `display-p3` for dst. Then, correct me if I misunderstood the code path, I do see `CreateExternalTexture` ([Chromium src](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/webgpu/external_texture_helper.cc;l=361;drc=42f0b9c4f60f67676d9096b3e94df9fd47f568f4;bpv=1;bpt=1)) 1-copy path supports `bt2020` & PQ transfer. Therefore, the result coming out of `textureSampleBaseClampToEdge()` will be color-converted (to srgb) and gamma corrected srgb/p3 value.

**Reply by shaoboyan** [reply, marked as answer]
I think this discussion is more like browser implementation details. Would you mind to file an chromium issue and we could continue discussion there?

>If textureSampleBaseClampToEdge is not proper choice for now,

One possible workaround could be "Using textureSampleBaseClampToEdge()" and make src_color_space and dst_color_space keep the same to avoid color space conversion. In this situation "textureSampleBaseClampToEdge()" should only do YUV->RGB conversion and you could apply HDR process based on this (But this needs display-p3 supported, which I think we could in implementation).

**Reply by kedanielwu** [reply]
Of course! I agree this is not a spec related question at current stage. I will file an Chromium issue later to continue the discussion, thanks for the help guys.

**Reply by kedanielwu** [reply]
@shaoboyan chromium issue link: https://bugs.chromium.org/p/chromium/issues/detail?id=1503556

sry I didn't see any option to tag the issue to Blink->WebGPU, currently it is in media category.
